src/mono/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.xml

```diff
@@ -633,6 +633,14 @@
 			<method name="get_BaseDirectory"/>
 		</type>
 
+		<type fullname="System.AppDomain">
+			<method name="DoUnhandledException"/>
+		</type>
+
+		<type fullname="System.Diagnostics.Debugger">
+			<method name="Mono_UnhandledException"/>
```

Comment:
If Xamarin Android is calling this, it should be designed and exposed as proper public API.

Reply:
Android *has* been calling this method for over 10 years: https://github.com/xamarin/monodroid/commit/3e9de5a51b

Other related reading:

   * https://github.com/xamarin/monodroid/commit/b0f85970102d43bab9cd860a8e8884d136d766b3
   * https://github.com/xamarin/monodroid/commit/12a012e00b4533d586ef31ced33351b63c9de883

Yes, it *should* be designed, exposed, and tested.  That said, this is what we *had*, and what *worked*.

Reply:
We already have `mono_unhandled_exception` which is part of the embedding API and should do the same thing. Android could also switch to using it instead.

Reply:
@BrzVlad : which raises a different question: `mono_unhandled_exception()` (added in 2002 via mono/mono@ce9497acaeddb9191876aaf09ec7dcead0a02ec5) predates xamarin/monodroid@3e9de5a (2010), so why wasn't that used in the first place?

*A* difference between the two is debugger interaction: `Debugger::Mono_UnhandledException`, in legacy, hits `debugger_agent_unhandled_exception()`:

  * https://github.com/mono/mono/blob/78c2b8da29881eccda4c27ca7b2b7e9ed451c55f/mono/mini/mini-runtime.c#L4680-L4683
  * https://github.com/mono/mono/blob/78c2b8da29881eccda4c27ca7b2b7e9ed451c55f/mono/mini/mini-runtime.c#L4307-L4318
  * https://github.com/mono/mono/blob/78c2b8da29881eccda4c27ca7b2b7e9ed451c55f/mono/mini/debugger-agent.c#L10288
  * https://github.com/mono/mono/blob/78c2b8da29881eccda4c27ca7b2b7e9ed451c55f/mono/mini/debugger-agent.c#L5125-L5143

Meanwhile, in `mono_unhandled_exception()`: https://github.com/mono/mono/blob/78c2b8da29881eccda4c27ca7b2b7e9ed451c55f/mono/metadata/object.c#L5105-L5211

I don't see any calls to `mini_get_dbg_callbacks()`, much less `mini_get_dbg_callbacks()->unhandled_exception`.

Calling `mono_unhandled_exception()` would allow re-working the `AppDomain.DoUnhandledException()` codepath (below), but is otherwise unrelated to `Debugger.Mono_UnhandledException`.

Reply:
> Yes, it should be designed, exposed, and tested. That said, this is what we had, and what worked.

Right, so I would expect as part of this PR:

- Code comments that explain that this is a legacy private reflection hack, with links where Xamarin Android calls it, so that the next person looking at this code won't delete it as dead code again.

- API proposal issue should be opened to create proper public managed API for this that Xamarin Android can switch to.

Reply:
@jonpryor Yeah, I was actually referring to the `DoUnhandledException` callback. I think it makes sense for `mono_unhandled_exception` to also call into the debugger. @thaystg seems to have looked into removing the `Debugger.Mono_UnhandledException` API altogether, maybe she has some ideas.

Anyhow. This is a regression on android on the new netcore profile and I'm not sure it is worthwhile to bother refactoring it so late in the release cycle. We could keep the legacy code for this release, and remove these API's for next release.

Reply:
I think this is the best approach to fix this regression, as I was working on implementing other things I couldn't look before preview 7.
I was planning to look at this, during this week, but maybe you are right it's too late for this release, we could only remove it on next release.

Reply:
@jonpryor , you don't see `mini_get_dbg_callbacks()->unhandled_exception` because with componentize work this was renamed to: `mono_component_debugger ()->unhandled_exception`, and it's still in the code:
https://github.com/dotnet/runtime/blob/c8f9a24ca0ebf6d9593e00ccbc7353510a0e0213/src/mono/mono/mini/mini-runtime.c#L4611

Reply:
@thaystg : the `Debugger::Mono_UnhandledException_internal` internal call is registered, but it doesn't exist: https://github.com/dotnet/runtime/blob/c8f9a24ca0ebf6d9593e00ccbc7353510a0e0213/src/mono/System.Private.CoreLib/src/System/Diagnostics/Debugger.cs

Hence the [change below](https://github.com/dotnet/runtime/pull/55904/files#diff-eaf0e5e77ad73ad572e8652891724341f4e7a6a27d889b12ef8f22043057bbe2R45), which re-introduces the declaration for the `Debugger::Mono_UnhandledException` internal call to `Debugger.cs`.